src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -84,19 +43,151 @@ BoundExpression TranslatePattern(BoundExpression input, BoundPattern pattern)
+        {
+            Debug.Assert(pattern.IsVar || pattern.LocalSymbol.Type == pattern.DeclaredType.Type);
+            if (pattern.IsVar)
+            {
+                Debug.Assert(input.Type == pattern.LocalSymbol.Type);
+                var assignment = _factory.AssignmentExpression(_factory.Local(pattern.LocalSymbol), input);
+                var result = _factory.Literal(true);
+                return _factory.Sequence(assignment, result);
+            }
+
+            return MakeDeclarationPattern(pattern.Syntax, input, pattern.LocalSymbol);
+        }
+
+        private BoundExpression LowerRecursivePattern(BoundRecursivePattern recursivePattern, BoundExpression input)
+        {
+            // cast the input to the argument type of 'operator is'.
+            var temp = _factory.SynthesizedLocal(recursivePattern.IsOperator.Parameters[0].Type, recursivePattern.Syntax);
+            var test = MakeDeclarationPattern(recursivePattern.Syntax, input, temp);
+
+            // prepare arguments to call 'operator is'
+            var arguments = ArrayBuilder<BoundExpression>.GetInstance();
+            var argumentTemps = ArrayBuilder<LocalSymbol>.GetInstance();
+            var argumentRefKinds = ArrayBuilder<RefKind>.GetInstance();
+            arguments.Add(_factory.Local(temp));
+            argumentRefKinds.Add(RefKind.None);
+            foreach (var p in recursivePattern.IsOperator.Parameters)
+            {
+                if (p.Ordinal == 0) continue;
+                var argumentTemp = _factory.SynthesizedLocal(p.Type, recursivePattern.Patterns[p.Ordinal - 1].Syntax);
+                argumentTemps.Add(argumentTemp);
+                argumentRefKinds.Add(RefKind.Out);
+                arguments.Add(_factory.Local(argumentTemp));
+            }
+
+            // invoke 'operator is' and use its result
+            BoundExpression invoke = _factory.Call(null, recursivePattern.IsOperator, argumentRefKinds.ToImmutableAndFree(), arguments.ToImmutableAndFree());
+            if (recursivePattern.IsOperator.ReturnsVoid)
+            {
+                // if the user-defined operator is irrefutable, we treat it as returning true
+                invoke = _factory.Sequence(invoke, _factory.Literal(true));
+            }
+
+            test = _factory.Sequence(temp, LogicalAnd(test, invoke));
+
+            // handle each of the nested patterns.
+            for (int i = 0; i < recursivePattern.Patterns.Length; i++)
+            {
+                var recursiveTest = LowerPattern(recursivePattern.Patterns[i], _factory.Local(argumentTemps[i]));
+                test = LogicalAnd(test, recursiveTest);
+            }
+
+            return _factory.Sequence(argumentTemps.ToImmutableAndFree(), test);
+        }
+
+        /// <summary>
+        /// Produce a 'logical and' operator that is clearly irrefutable when it can be.
+        /// </summary>
+        BoundExpression LogicalAnd(BoundExpression left, BoundExpression right)
```

Comment:
No. It is designed specifically for the needs of the following method, and both are designed just to be used for pattern-matching.